### ISSUE TITLE
gtk: window border is not removed if `window-decoration` is `false`

### ISSUE THREAD

**Issue by echasnovski** [ghostty-org/ghostty#2023]
On current `main` (4cf62029) window border is not removed with `window-decoration=false` but it is removed with `gtk-titlebar=false`. 
I've bisected the behavior change to 2d7baaa7 (i.e. #1950).

As [documentation says](https://github.com/ghostty-org/ghostty/blob/4cf620296fefd36dcbb1b761fa67c9e858948d77/src/config/Config.zig#L701-L702) that both titlebar and borders should be disabled when `window-decoration=false`, I am not sure if this is an issue with documentation or code.

OS: Linux
Window manager: i3

Screenshots:

`ghostty` (with default config):

![screenshot_2024-08-01_12:15:56](https://github.com/user-attachments/assets/33107e1e-e8f8-49aa-a875-b9e15ad89683)

`ghostty --window-decoration=false`:

![screenshot_2024-08-01_12:16:16](https://github.com/user-attachments/assets/edd5513d-ff25-4b9c-bd6c-06eaec137285)

`ghostty --gtk-titlebar=false`:

![screenshot_2024-08-01_12:16:34](https://github.com/user-attachments/assets/343a275e-2df5-40ce-95bb-572204b10d13)

**Comment by tjbarber** [comment]
Confirmed that this is still happening, but they do go away with:

```
gtk-titlebar=false
window-decoration=false
```

**Comment by davidbudzynski** [comment]
I'm having similar issue using Arch and Qtile WM. It looks like `window-decoration = false` yields removes decorations but rounded corners stay like so: 

![Image](https://github.com/user-attachments/assets/0f849dee-f315-4259-b0d0-f5cc80b66ac9)

if I comment out `window-decoration = false` and add `gtk-titlebar = false`, this solves my issue with rounded corners being present.

![Image](https://github.com/user-attachments/assets/03da62ca-f7fc-4dea-86a4-87a07ee57d69)

I'm assuming that `gtk-titlebar` is Linux specific or Linux with a WM specific and 
`window-decoration` is irrelevant?

**Comment by treeman** [comment]
As of today (`1.0.2-main+60611b8a`) I can no longer disable borders with `window-decoration` nor `gtk-titlebar`:

```
window-decoration = false
gtk-titlebar = false
```

![Image](https://github.com/user-attachments/assets/4db51fd5-1f0a-495a-b383-56188f39cd83)

`gtk-titlebar` doesn't seem to have any effect anymore.

**Comment by davidbudzynski** [comment]
Same here actually.

**Comment by jcollie** [comment]
See #4224

**Comment by treeman** [comment]
> See [#4224](https://github.com/ghostty-org/ghostty/discussions/4224)

Thank you, `gtk-adwaita = false` made the borders go away again.

**Comment by Gelmo** [comment]
Small problem. With `gtk-adwaita = false`, the tabs are no longer respecting the window theme. Config:
```
theme = DraculaFix
window-theme = ghostty
gtk-titlebar = false
gtk-adwaita = false
```
Screenshot:
![Image](https://github.com/user-attachments/assets/88ff7548-e4ed-4763-85d1-e868b74ce8c5)

Description:

The top window shows `gtk-adwaita = true`, and the bottom window shows `gtk-adwaita = false`. I'd like to have the lack of border from the bottom window, with the colors used in the top window. I actually prefer the styling of the non-Adwaita tab bar, but wish it was the same colors as specified for the Ghostty theme.

I think the following should all be configurable by the user:
- No title bar (both windows shown above)
- No border (the bottom window shown above)
- Properly themed tabs regardless of whether or not Adwaita is used

Ideal config:
```
theme = DraculaFix
window-theme = ghostty
gtk-titlebar = false
gtk-border = false
```
Or
```
theme = DraculaFix
window-theme = ghostty
gtk-titlebar = false
gtk-border = false
gtk-adwaita = false
```
The only difference in the above configs would be the layout/rounding/highlighting of the tab bar (Adwaita layout vs vanilla GTK layout), with both using the same colors.

**Comment by tristan957** [comment]
> Small problem. With gtk-adwaita = false, the tabs are no longer respecting the window theme.

Yes, this is expected behavior, and is documented as such. Please read the documentation.

**Comment by Gelmo** [comment]
> Yes, this is expected behavior, and is documented as such. Please read the documentation.

Right, I have done so. I'm posting because my desired behavior is not currently supported, and it's related to this issue. As discussed in this issue and https://github.com/ghostty-org/ghostty/discussions/4224 the gtk window border is shown unless you set `gtk-adwaita = false`.

I'm stating that setting `gtk-adwaita = false` results in other changes, as well, which are undesirable to this user (me). And therefore, having a way to disable the border while respecting the used theme could be beneficial.

I wanted to post my comment in response to this - https://github.com/ghostty-org/ghostty/discussions/4224#discussioncomment-11711031 - However, with that discussion being listed as "Answered" I feel that it may not get attention from the appropriate people; that's why I posted here.

I doubt I am the only person who wants to be able to remove the border while also having their set GTK theme respected. Maybe I am now, with Ghostty being recently made public, but over time there will be others with the same gripe.

I use a non-Adwaita GTK theme. I don't think it's unreasonable to expect to be able to use my set GTK theme without being forced onto another theme (Adwaita) that doesn't fit the rest of my environment.

Should I post this in the other discussion? Open a new discussion? Give up on any hope of Ghostty being able to be configured as desired and go back to another terminal emulator?

**Comment by Gelmo** [comment]
Here's a screenshot of another GTK application that respects my set theme

![Image](https://github.com/user-attachments/assets/e7418b4d-9da4-44ac-a719-cf949f16072c)

Just to confirm that the issue is unrelated to my set theme.

**Comment by tristan957** [comment]
> I'm stating that setting gtk-adwaita = false results in other changes, as well, which are undesirable to this user (me). And therefore, having a way to disable the border while respecting the used theme could be beneficial.

Using Adwaita has pros and cons, and is something that every user must balance for their own sake. Reworking the Adwaita theme to meet expectations that it is not meant for is not something that I would advocate for within Ghostty.

**Comment by Gelmo** [comment]
> Using Adwaita has pros and cons, and is something that every user must balance for their own sake. Reworking the Adwaita theme to meet expectations that it is not meant for is not something that I would advocate for within Ghostty.

Yes, I agree that adwaita behavior shouldn't be modified. I'm saying that when adwaita is disabled, it should use the globally-set GTK theme.

**Comment by jcollie** [comment]
We're going to explore the use of some Wayland protocols to enable server-side decorations if we're not running on Gnome. So it may be possible to keep Adwaita _and_ have server-side decorations. No promises, this is at the concept stage and very experimental and probably won't make 1.0.2.

**Comment by tristan957** [comment]
> Yes, I agree that adwaita behavior shouldn't be modified. I'm saying that when adwaita is disabled, it should use the globally-set GTK theme, and in my case, it doesn't unless the theme is installed locally for the user (which is not necessary for other GTK applications that I utilize)

We are not doing anything that would preclude GTK from picking up your theme if adwaita is disabled.

**Comment by Gelmo** [comment]
> We're going to explore the use of some Wayland protocols to enable server-side decorations if we're not running on Gnome.

What about people using X11?

**Comment by Gelmo** [comment]
> We are not doing anything that would preclude GTK from picking up your theme if adwaita is disabled.

That's what I would have assumed, but it doesn't explain why other GTK apps didn't require my theme to be installed per-user

**Comment by mitchellh** [comment]
> What about people using X11?

We have experienced volunteers doing good work pursuing Wayland improvements and don’t presently have the same enthusiasm for X11. I’m not against it at all, it’s just the present state of contribution.

**Comment by JenChampagne** [comment]
For what it's worth, I'm on X11 and was able to hide borders with the following config when I was on commit `037de64e`:

```
gtk-adwaita=false
gtk-titlebar=false
window-decoration=false
```

I just upgraded to commit `ff9414d9` and borders are back despite those configs. I was able to narrow it down to [commit `c4ece2a`](https://github.com/ghostty-org/ghostty/commit/c4ece2a141d77f7fb5560daec46b60972faf3f92), so I'll stay on the prior commit for now.

I don't have the ability to help with code at the moment but I'm happy to be a test subject at least 😆 

Some of my system details in case it's helpful:
OS: Linux hostname 6.5.0-0.deb12.4-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.5.10-1~bpo12+1 (2023-11-23) x86_64 GNU/Linux
X.Org version: 1.21.1.7
i3: 4.22 (2023-01-02)

**Comment by treeman** [comment]
With Ghostty 1.1.0 the borders are still present.

With `window-decoration = none` I still get a small border wheile on commit `1baf8928` I don't get this border (`1baf8928` on the left without a border, and 1.1.0 on the right with a small dark blue border):

![Image](https://github.com/user-attachments/assets/e932963c-c715-4423-bb6e-fb243250f04c)

If I add `gtk-adwaita = false` then i get the same border as in the second image of OP.

**Comment by treeman** [comment]
The small extra border I wrote about above is still there after the latest fix (small blue border to the right):

![](https://private-user-images.githubusercontent.com/162027/408459533-e932963c-c715-4423-bb6e-fb243250f04c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg3ODM1NTAsIm5iZiI6MTczODc4MzI1MCwicGF0aCI6Ii8xNjIwMjcvNDA4NDU5NTMzLWU5MzI5NjNjLWM3MTUtNDQyMy1iYjZlLWZiMjQzMjUwZjA0Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNVQxOTIwNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mODI1MDQ1YTI5YzE5NTIzY2NhNzA2Y2ZjZjFmMTk1Y2M2ZWZiNWMwYzM0ZGNkYWViYjFiZjQyZDdjNTVjYTk1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.RNCVO9Os3S8Sm69UZHYK2eLq6wEtkhrzXyykh2OmsD8)

**Comment by CatalinVoineag** [comment]
I still have this issue even after the fix was merged.

I managed to fix mine by adding custom CSS to my ghostty config

```

# ghostty/config file
window-decoration = false
gtk-custom-css = ./custom.css


# custom.css file
window {
  padding: 0
}

```

**Comment by RileyMathews** [comment]
Just chiming in here thanks to @CatalinVoineag for the reference. That also helped me remove the border that seemed to crop back in after updating. I know Ghostty is trying to cater to a very wide audience but just to give my preferences as someone who enjoys using a very minimal environment driven by windows managers like i3 and qtile having easily configurable options that strip away as much extraneous UI as possible would be amazing. But I understand if that isn't the direction Ghostty as a whole is going.

**Comment by tristan957** [comment]
> Just chiming in here thanks to [@CatalinVoineag](https://github.com/CatalinVoineag) for the reference. That also helped me remove the border that seemed to crop back in after updating. I know Ghostty is trying to cater to a very wide audience but just to give my preferences as someone who enjoys using a very minimal environment driven by windows managers like i3 and qtile having easily configurable options that strip away as much extraneous UI as possible would be amazing. But I understand if that isn't the direction Ghostty as a whole is going.

You should investigate https://github.com/gabydd/wraith. It is going to be the better choice for WM users.
